enhancements/security/scc-csi-volumes.md

```diff
@@ -0,0 +1,234 @@
+* Enable SCCs to limit [generic ephemeral volumes](https://kubernetes.io/docs/concepts/storage/ephemeral-volumes/#generic-ephemeral-volumes)
+
+## Proposal
+
+### User Stories
+
+As an OpenShift cluster admin
+I want to control which CSI drivers can be used to mount csi volumes
+So that my workloads only use trusted CSI drivers to access content
+
+As an application developer
+I want to use the secret store CSI driver to mount csi volumes
+So that I can access sealed secrets in my workloads
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### Safe for Consumption Annotation
+
+CSI drivers that provide inline CSI volumes must mark themselves as safe for consumption.
+To mark a CSI driver safe, admins or an operator will need to add the `security.openshift.io/allow-csi-volumes-restricted` annotation to the driver's `CSIDriver` object:
+
+```yaml
+apiVersion: storage.k8s.io/v1
+kind: CSIDriver
+annotations:
+  security.openshift.io/allow-csi-volumes-restricted: "true"
+  ...
+metadata:
+  name: mycsi.dev
+  ...
+```
+
+#### Pod admission
+
+Pod that mount `csi` volumes cannot be admitted unless the chosen SCC allows `csi` volumes.
+In addition, the driver for the volume mount must have the `security.openshift.io/allow-csi-volumes-restricted` annotation.
+
+#### Allow CSI volumes in restricted namespaces
+
+Ephemeral CSI drivers will be promoted to the list of "safe" volume mounts, and be considered equivalent to a Secret or ConfigMap from a safety/restriction perspective.
+CSI volumes will be added to the list of allowed volume types in the `restricted` SCC.
+All default SCCs installed by OpenShift with less restrictions will likewise allow `csi` volumes.
+When SCCs are ordered, the `csi` volume type will count as a "trivial" volume mount.
+
+### Risks and Mitigations
+
+**Risk:**
+Default security context constraints will allow any CSI driver to be used.
+
+*Mitigation:*
+Only the CSI drivers that are marked safe by an operator or admin will be allowed to mount `csi` volumes.
+
+**Risk:**
+Checking the annotation on the CSI driver will not be backwards compatible.
+
+*Mitigation:*
+
+Documentation for storage and SCCs will need to communicate that inline CSI volume drivers will need to add the `security.openshift.io/allow-csi-volumes-restricted` annotation to work on OpenShift.
+A CSI driver that mounts inline CSI volumes should be safe for any user to use on OpenShift, or it should not be allowed at all.
+
+**Risk:**
+During upgrade, any pod can use any CSI driver due to version skew between the default SCCs and the SCC admission logic in openshift-apiserver.
+
+*Mitigation:*
+The upgrade test suites must ensure that pods cannot be created with `csi` volumes in a restricted namespace.
+
+## Design Details
+
+### Open Questions [optional]
+
+1. Should we allow the privileged SCC to mount any CSI volume (via an `AllowAllCSIVolumeDrivers` flag)?
+   This has implications for OpenShift builds, which currently uses the `privleged` SCC.
```

Comment:
If such an admin "get of jail" flag is added, then we should consider a separate SCC for builds. We will likely want this anyway as we enable unprivileged builds.